packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "GoogleMapHeatmapController.h"
+#import "JsonConversions.h"
+
+@implementation FLTGoogleMapHeatmapController {
+  GMUHeatmapTileLayer* _heatmapLayer;
+  GMSMapView* _mapView;
+}
+- (instancetype)initHeatmapWithHeatmapId:(NSString*)heatmapId mapView:(GMSMapView*)mapView {
+  self = [super init];
+  if (self) {
+    _heatmapLayer = [[GMUHeatmapTileLayer alloc] init];
+    _heatmapLayer.map = mapView;
+    _mapView = mapView;
+    _heatmapId = heatmapId;
+  }
+  return self;
+}
+
+- (void)removeHeatmap {
+  _heatmapLayer.map = nil;
+}
+
+#pragma mark - FLTGoogleMapHeatmapOptionsSink methods
+
+- (void)setVisible:(BOOL)visible {
+  _heatmapLayer.map = visible ? _mapView : nil;
+}
+- (void)setRadius:(NSUInteger)radius {
+  [_heatmapLayer setRadius:radius];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setGradient:(GMUGradient*)gradient {
+  [_heatmapLayer setGradient:gradient];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setPoints:(NSArray<GMUWeightedLatLng*>*)points {
+  [_heatmapLayer setWeightedData:points];
+  [_heatmapLayer clearTileCache];
+}
+- (void)setOpacity:(double)opacity {
+  _heatmapLayer.opacity = opacity;
+  [_heatmapLayer clearTileCache];
+}
+
+@end
+
+static int ToInt(NSNumber* data) { return [FLTGoogleMapJsonConversions toInt:data]; }
+
+static double ToDouble(NSNumber* data) { return [FLTGoogleMapJsonConversions toDouble:data]; }
+
+static BOOL ToBool(NSNumber* data) { return [FLTGoogleMapJsonConversions toBool:data]; }
+
+static NSArray<GMUWeightedLatLng*>* ToPoints(NSArray* data) {
+  NSMutableArray* points = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    NSNumber* latitude = data[i][0][0];
+    NSNumber* longitude = data[i][0][1];
+    NSNumber* intensity = data[i][1];
+    GMUWeightedLatLng* weightedPoint = [[GMUWeightedLatLng alloc]
+        initWithCoordinate:CLLocationCoordinate2DMake(
+                               [FLTGoogleMapJsonConversions toDouble:latitude],
+                               [FLTGoogleMapJsonConversions toDouble:longitude])
+                 intensity:[FLTGoogleMapJsonConversions toFloat:intensity]];
+    [points addObject:weightedPoint];
+  }
+
+  return points;
+}
+
+static NSArray<UIColor*>* ToColors(NSArray* data) {
+  NSMutableArray* colors = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    UIColor* color = data[i];
+    [colors addObject:color];
+  }
+
+  return colors;
+}
+
+static NSArray<NSNumber*>* ToStartsPoints(NSArray* data) {
+  NSMutableArray* startPoints = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    NSNumber* startPoint = data[i];
+    [startPoints addObject:startPoint];
+  }
+
+  return startPoints;
+}
```

Comment:
The method seems unnecessary. The method seems unnecessary. It doesn't seem to be doing anything with the data in the array. The method just did a deep copy from `data` to the return value.

Reply:
This method converts a JSON object (NSArray* data) to a array of NSNumber*. I am no iOS developer but it didn't seem to work without this method.